pkg/trie/trie_tools.go

```diff
@@ -110,6 +111,78 @@ func (s *Trie) get(root, key []byte, batch [][]byte, iBatch, height int) ([]byte
 	return s.get(lnode, key, batch, 2*iBatch+1, height-1)
 }
 
+// WalkResult contains the key and value obtained with a Walk() operation
+type WalkResult struct {
+	Value []byte
+	Key   []byte
+}
+
+// Walk finds all the trie stored values from left to right and calls callback.
+// If callback returns true, the walk will stop, else it will continue.
+func (s *Trie) Walk(callback func(*WalkResult) bool) error {
+	walkc := make(chan *WalkResult)
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	s.atomicUpdate = false
+	close := make(chan (bool), 1)
+	stop := false
+	wg := sync.WaitGroup{}
+	go func() {
+		for {
+			select {
+			case <-close:
+				break
```

Comment:
I tracked down the random bug you mentioned to this function. There are cases where the Walk() finishes iterating the tree and returns while the callback hasn't finished executing. So if calling Walk() 2 times in a row, the `i++` from the 1st walk can override the `i = 0` reset for the 2nd walk.

To prevent this, you need confirmation that the goroutine has finished executing before Walk() returns, by using a blocking channel that waits for the goroutine to return for example.

Also, this `break` should be a `return` otherwise only select is broken and not the `for` loop.

Using WaitGroup doesn't seem necessary here.

`close` channel can be renamed as it is a golang builtin

For example:
```
	s.lock.RLock()
	defer s.lock.RUnlock()
	s.atomicUpdate = false
	walkc := make(chan *WalkResult)
	exit := make(chan (bool))
	finishedWalk := make(chan (bool))
	stop := false
	go func() {
		for {
			select {
			case <-finishedWalk:
                                exit <- true
				return
			case value := <-walkc:
				if stop = callback(value); stop {
                                        // break and loop to case <- finishedWalk
					break
				}
			}
		}
	}()
	err := s.walk(walkc, &stop, s.Root, nil, 0, s.TrieHeight)
	finishedWalk <- true
        <- exit. // wait for goroutine to return
	return err

```